Feature fix quality metrics

roigcarlo opened the pull request:
This should fix #94 and fix #85. 

As a resume this PR:
* Corrects the problem with the `VolumeToAverageEdgeLength` quality function for tetrahedras
* Fixed a bug that was causing an error if tests were compiled with clang
* Completed the missing tests
* Extended the tests for tetrahedras so a non-unitary geometry is also tested
* Non implemented functions (regualrity and surface area) are now check that the excpetion is trown rather than the correct value.
* Fixed KRATOS_CHECK_EXCEPTION while being called multiple times inside the same test

roigcarlo committed:
Added non unitary triangle to the tests

roigcarlo committed:
Added missing test and checks

roigcarlo committed:
Incorrect values

roigcarlo committed:
Adding more compilers to this check (all of them actually)

roigcarlo committed:
Fixed face number, valume and area

roigcarlo committed:
Fixed VolumeToaverageEdgeLenght missing a ^3 and corrected normalization Factor

roigcarlo committed:
Corrected test value to reflec the changes in the quality function

roigcarlo committed:
Moving some var out of the macro so it can be called multiple times

roigcarlo committed:
Not implemented functions now have their checks set to KRATOS_CHECK_EXCEPTION

loumalouomega commented:
![](https://media.giphy.com/media/aLdiZJmmx4OVW/giphy.gif)

NICE WORK

pooyan-dadvand commented on kratos/testing/tester.cpp:
```diff
@@ -228,8 +228,7 @@ namespace Kratos
 
-#if defined(__GNUC__) &&( __GNUC__ < 4 || \
-              (__GNUC__ == 4 && (__GNUC_MINOR__ < 9))) 
+#if __cplusplus <= 199711L
```
Is this also visual studio compatible?

roigcarlo replied:
There was a problem in VS 2012 with this but since we already use 2015 it should be fine. I will test it again to be sure.

roigcarlo replied:
Looks like they still define this as 199711L.

The question is, since we now have C+11 as a requisite can't we just compile the regex code as is?

RiccardoRossi replied:
I think that this particular feature (regex) was still broken on relatively recent gccs, anyway Pooyan will remember more about this

pooyan-dadvand replied:
Exactly as @RiccardoRossi saying, gcc 4.8 which is one of the popular old one does not support it and for me this function wasn't enough for drop it

pooyan-dadvand commented on kratos/tests/geometries/test_geometry.h:
```diff
@@ -28,14 +28,15 @@
 namespace Kratos {
 namespace Testing {
 
+    static std::ofstream out("/dev/null");
```
Is this windows compatible?

roigcarlo replied:
Probably not. 
As we spoke, maybe I should just remove that ?

roigcarlo replied:
At the end it was working because `std::cout.rdbuf(out.rdbuf());` was evaluated as `NULL`, which is an accepted buffer which has the same effect as `/dev/null` in linux. As it happens to be what we want im just making `std::cout.rdbuf(nullptr);`

According to [this](http://stackoverflow.com/questions/14845446/how-redirect-stdostream-to-a-file-or-dev-null-in-c-and-linux) it also sets the `std::ios::badbit` which should prevent conversions in the output in case we eventually have any.

pooyan-dadvand replied:
Seems ok, but in long term I would like to know why the same mechanism is not working in the [tester](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/testing/tester.cpp#L259)

pooyan-dadvand reviewed (commented):
I think there are some compatibility issues for windows to be checked

roigcarlo committed:
Changind /dev/null to nullptr

roigcarlo committed:
Restored old macro and added explicit case for clang

roigcarlo committed:
Fixing length

roigcarlo committed:
Sqrt is not constexpr

roigcarlo committed:
Add negative volume tetrahedra in tests

roigcarlo committed:
Added test for length

roigcarlo committed:
clarify what is that constant

roigcarlo committed:
Clarify the method used to calculate the volume to avoid another volume pandemonium

roigcarlo committed:
Extended tests for the inverted case

roigcarlo committed:
Fixed bug while calculating the circumradius

roigcarlo committed:
Merge branch 'feature-fix-quality-metrics' of https://github.com/KratosMultiphysics/Kratos into feature-fix-quality-metrics

roigcarlo committed:
Typos

roigcarlo committed:
length -> characteristic length

roigcarlo commented:
@pooyan-dadvand I've made some changes since the last time:
* The tester macro should now test the clang case apart
* Fixed the length returning `-nan` from #126. Notice that it uses the norm of the volume so it will be always positive
* I've corrected a problem in the determinants while calculating the circumsphere
* I added an inverted tetrahedra to the tests and extended the tests so it is checked as well, returning negative values in those who use volume as a metric (which was already working btw).

However, none of the quality functions were using the `Length()` method from the geometry, so the problems in #126 may still persist...

pooyan-dadvand commented on kratos/geometries/tetrahedra_3d_4.h:
```diff
@@ -506,9 +500,9 @@ template<class TPointType> class Tetrahedra3D4 : public Geometry<TPointType>
-      double detJMX = s10 * y20 * z30 + y10 * z20 * s30 + z10 * s20 * y30 - s30 * y20 * z30 - y30 * z20 * s10 - z30 * s20 * y10;
-      double detJMY = s10 * x20 * z30 + x10 * z20 * s30 + z10 * s20 * x30 - s30 * x20 * z30 - x30 * z20 * s10 - z30 * s20 * x10;
-      double detJMZ = s10 * x20 * y30 + x10 * y20 * s30 + y10 * s20 * x30 - s30 * x20 * y30 - x30 * y20 * s10 - y30 * s20 * x10;
+      double detJMX = s10 * y20 * z30 + y10 * z20 * s30 + z10 * s20 * y30 - s30 * y20 * z10 - y30 * z20 * s10 - z30 * s20 * y10;
```
I should not change this! The volume results to be correct after my issue. Why you want to change it?

roigcarlo replied:
This is not the volume function. This calculates the circumradius of the tetrahedron for the quality function that need them. It had an error in one of the components of a diagonal that calculates another determinant. See ac707a565561248d556e0121634d586fe669350b. 

But I can rewrite the whole matrix so it looks like the volume if you want.

pooyan-dadvand replied:
Now I see. So I aree with changes.

pooyan-dadvand reviewed (changes requested):
I would not let change the Volume method unless you have a very strong reason because it will break a lot of code.

roigcarlo commented:
Can we merge?
